tools/postgres/dbtool.js

```diff
@@ -0,0 +1,447 @@
+      fs
+        .readFileSync(path.join(dumpDir, filename), { encoding: "utf-8" })
+        .replace(/,$/gm, "")
+        .replace(/\\r/gm, "  ")
+        .split("\n")
+        .sort()
+        .join("\n"),
+    );
+  }
+}
+
+function dumpSchemaViaSqlFiles(filePaths) {
+  const tmpDbName = `webknossos_tmp_${Date.now()}`;
+  const tmpSchemaDir = fs.mkdtempSync("temp-webknossos-schema-");
+
+  try {
+    // Create tmp database
+    safePsqlSpawn([PG_CONFIG.urlWithoutDatabase, "-c", `CREATE DATABASE ${tmpDbName}`]);
+
+    const urlWithDatabase = new URL(PG_CONFIG.urlWithoutDatabase);
+    urlWithDatabase.pathname = "/" + tmpDbName;
+    // Load schema into tmp database
+    safePsqlSpawn([
+      urlWithDatabase.toString(),
+      "-v",
+      "ON_ERROR_STOP=ON",
+      "-q",
+      ...filePaths.flatMap((filePath) => ["-f", filePath]),
+    ]);
+
+    // Dump schema into diffable files
+    dumpSchema(urlWithDatabase.toString(), tmpSchemaDir, true);
+    return tmpSchemaDir;
+  } finally {
+    safePsqlSpawn([PG_CONFIG.urlWithoutDatabase, "-c", `DROP DATABASE ${tmpDbName}`]);
+  }
+}
+
+function checkDbSchema() {
+  const dbDumpDir = fs.mkdtempSync("temp-webknossos-schema-");
+  let schemaDumpDir = null;
+  try {
+    dumpSchema(PG_CONFIG.url, dbDumpDir, true);
+    schemaDumpDir = dumpSchemaViaSqlFiles([schemaPath]);
+
+    cleanSchemaDump(dbDumpDir);
+    cleanSchemaDump(schemaDumpDir);
+
+    try {
+      safeSpawn("diff", ["--strip-trailing-cr", "-r", dbDumpDir, schemaDumpDir]);
+    } catch (err) {
+      throw new Error(`Database schema is not up-to-date:\n${err.stdout}`);
+    }
+  } finally {
+    if (fs.existsSync(dbDumpDir)) {
+      fs.rmSync(dbDumpDir, { recursive: true, force: true });
+    }
+    if (fs.existsSync(schemaDumpDir)) {
+      fs.rmSync(schemaDumpDir, { recursive: true, force: true });
+    }
+  }
+  console.log("✨✨ Database schema is up-to-date");
+}
+
+function checkEvolutionsSchema() {
+  let evolutionsDumpDir = null;
+  let schemaDumpDir = null;
+  try {
+    evolutionsDumpDir = dumpSchemaViaSqlFiles(
+      fs
+        .readdirSync(evolutionsPath)
+        .filter((filename) => filename.endsWith(".sql"))
+        .map((filename) => path.join(evolutionsPath, filename))
+        .sort(),
+    );
+    schemaDumpDir = dumpSchemaViaSqlFiles([schemaPath]);
+
+    cleanSchemaDump(evolutionsDumpDir);
+    cleanSchemaDump(schemaDumpDir);
+
+    try {
+      safeSpawn("diff", ["--strip-trailing-cr", "-r", evolutionsDumpDir, schemaDumpDir]);
+    } catch (err) {
+      throw new Error(`Evolutions do not match with schema file:\n${err.stdout}`);
+    }
+  } finally {
+    if (fs.existsSync(evolutionsDumpDir)) {
+      fs.rmSync(evolutionsDumpDir, { recursive: true, force: true });
+    }
+    if (fs.existsSync(schemaDumpDir)) {
+      fs.rmSync(schemaDumpDir, { recursive: true, force: true });
+    }
+  }
+  console.log("✨✨ Evolutions match with schema file");
+}
+
+function applyEvolutions() {
```

Comment:
I don’t think this is used anywhere, do you think we should keep it anyway?

Reply:
let's keep it